core/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java

```diff
@@ -0,0 +1,376 @@
+                            }
+
+                            @Override
+                            public void handleException(TransportException exp) {
+                                onFailure(idx, node.id(), exp);
+                            }
+
+                            @Override
+                            public String executor() {
+                                return ThreadPool.Names.SAME;
+                            }
+                        });
+                    }
+                } catch (Throwable t) {
+                    onFailure(idx, nodeId, t);
+                }
+            }
+        }
+
+        private void onOperation(int idx, NodeTasksResponse nodeResponse) {
+            responses.set(idx, nodeResponse);
+            if (counter.incrementAndGet() == responses.length()) {
+                finishHim();
+            }
+        }
+
+        private void onFailure(int idx, String nodeId, Throwable t) {
+            if (logger.isDebugEnabled() && !(t instanceof NodeShouldNotConnectException)) {
+                logger.debug("failed to execute on node [{}]", t, nodeId);
+            }
+            if (accumulateExceptions()) {
+                responses.set(idx, new FailedNodeException(nodeId, "Failed node [" + nodeId + "]", t));
+            }
+            if (counter.incrementAndGet() == responses.length()) {
+                finishHim();
+            }
+        }
+
+        private void finishHim() {
+            TasksResponse finalResponse;
+            try {
+                finalResponse = newResponse(request, responses);
+            } catch (Throwable t) {
+                logger.debug("failed to combine responses from nodes", t);
+                listener.onFailure(t);
+                return;
+            }
+            listener.onResponse(finalResponse);
+        }
+    }
+
+    class NodeTransportHandler implements TransportRequestHandler<NodeTaskRequest> {
+
+        @Override
+        public void messageReceived(final NodeTaskRequest request, final TransportChannel channel) throws Exception {
+            channel.sendResponse(nodeOperation(request));
+        }
+    }
+
+
+    private class NodeTaskRequest extends ChildTaskRequest {
+        private TasksRequest tasksRequest;
+
+        protected NodeTaskRequest() {
+            super();
+        }
+
+        protected NodeTaskRequest(TasksRequest tasksRequest) {
+            super(tasksRequest);
+            this.tasksRequest = tasksRequest;
+        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            super.readFrom(in);
+            tasksRequest = requestSupplier.get();
+            tasksRequest.readFrom(in);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            tasksRequest.writeTo(out);
+        }
+    }
+
+    private class NodeTasksResponse extends TransportResponse {
```

Comment:
That would significantly complicate things. `NodeTasksResponse` needs access to `TransportTasksAction` because it is using `TransportTasksAction#readTaskResponse` to read tasks.